### PR TITLE
Fix gRPC instrumentation on recent kernels

### PR DESCRIPTION
**Pull request by edeNFed**
In recent kernel versions, the eBPF verifier checks that a valid number of bytes (positive and not overflow) are being written/read from memory.
This PR adds additional bound checks so gRPC instrumentation now passes verification.

Tested on Ubuntu 20.02, x86, running in DigitalOcean.

- Closes https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/78
- Closes https://github.com/open-telemetry/opentelemetry-go-instrumentation/issues/141

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,10 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
 
 ## [Unreleased]
 
+### Changed
```
Comment (MrAlias) [inline CHANGELOG.md:11]: nit

```suggestion
### Fixed
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,10 @@ OpenTelemetry Go Automatic Instrumentation adheres to [Semantic Versioning](http
...
+
+- Fix gRPC instrumentation on recent kernels ([#150](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/150))
```
Comment (MrAlias) [inline CHANGELOG.md:13]:
```suggestion
- Fix gRPC instrumentation memory access issue on newer kernels. ([#150](https://github.com/open-telemetry/opentelemetry-go-instrumentation/pull/150))
```

**Review by MrAlias — APPROVED** [review]

File: include/alloc.h [hunk 3]
```diff
@@ -65,6 +66,22 @@ static __always_inline u64 get_area_end(u64 start)
...
+{
+    if (num < min)
+    {
+        return min;
+    }
+    else if (num > max)
+    {
+        return max;
+    }
+    else
```
Comment (pdelewski) [inline include/alloc.h:79]: This else seems to be redundant

File: include/alloc.h [hunk 4]
```diff
@@ -83,6 +100,7 @@ static __always_inline void *write_target_data(void *data, s32 size)
     }
 
     void *target = (void *)start;
+    size = bound_number(size, 1, MAX_BUFFER_SIZE);
```
Comment (pdelewski) [inline include/alloc.h:103]: `const size`?

**Comment by edeNFed** [comment]
@vreynolds thanks for trying it out! I tested it with the [emojivoto application](https://github.com/open-telemetry/opentelemetry-go-instrumentation/tree/main/docs/getting-started/emojivoto) and sent the traces to Jaeger which looked good (correct spans + context propagation

**Review by MikeGoldsmith — APPROVED** [review]

**Pull request merged**